### ISSUE TITLE
OS X: recursion in the dylib-unhell.py function process_libraries can cause "maximum recursion depth exceeded"

### ISSUE THREAD

**Issue by mag01** [mpv-player/mpv#3316]
### mpv version and platform

MPV 0.18.0 release
OS X 10.11.6 Beta (15G24b)
### Reproduction steps

build MPV from sources
then use "TOOLS/osxbundle.py build/mpv"
### Expected behavior

the mpv.app package is created successfully
### Actual behavior

Python exception occurs durring the osxbundle.py run:
RuntimeError: maximum recursion depth exceeded while calling a Python object
### Log files

$ TOOLS/osxbundle.py build/mpv

```
Creating Mac OS X application bundle (version: 0.18.0)...
> copying bundle skeleton
> copying binary
> generating Info.plist
> bundling dependencies
Traceback (most recent call last):
  File "TOOLS/dylib-unhell.py", line 94, in <module>
    main()
  File "TOOLS/dylib-unhell.py", line 91, in main
    process_libraries(libs, binary)
  File "TOOLS/dylib-unhell.py", line 83, in process_libraries
    process_libraries(libs_dict, binary, ls)
...
  File "TOOLS/dylib-unhell.py", line 83, in process_libraries
    process_libraries(libs_dict, binary, ls)
  File "TOOLS/dylib-unhell.py", line 63, in process_libraries
    ls   = leafs(libs_dict, processed)
  File "TOOLS/dylib-unhell.py", line 48, in leafs
    processed = set(processed)
RuntimeError: maximum recursion depth exceeded while calling a Python object
done.
```

Output was cut down - there are about 1000 repetitions of:

```
File "TOOLS/dylib-unhell.py", line 83, in process_libraries
    process_libraries(libs_dict, binary, ls)
```

in the traceback.

Increasing the recursion limit ( sys.setrecursionlimit(limit) ) doesn't actually help.
The value where it stops failing as above is about 14200 in my case, but even though it stops failing as above then, the whole Python process will start crashing instead ( EXC_BAD_ACCESS (SIGSEGV) ).

So this will need to be rewritten to the iterative approach.

**Comment by pigoz** [comment]
Hi, could you paste the output of `otool -L build/mpv | grep -e '\t' | awk '{ print $1 }'` ?

**Comment by mag01** [comment]
This is the whole output:
```
/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
/System/Library/Frameworks/Security.framework/Versions/A/Security
/System/Library/Frameworks/OpenCL.framework/Versions/A/OpenCL
/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon
/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation
/System/Library/Frameworks/IOSurface.framework/Versions/A/IOSurface
/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa
/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit
/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL
/System/Library/Frameworks/QuartzCore.framework/Versions/A/QuartzCore
/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit
/System/Library/Frameworks/QTKit.framework/Versions/A/QTKit
/System/Library/Frameworks/CoreVideo.framework/Versions/A/CoreVideo
/System/Library/Frameworks/AVFoundation.framework/Versions/A/AVFoundation
/System/Library/Frameworks/CoreMedia.framework/Versions/A/CoreMedia
/System/Library/Frameworks/VideoToolbox.framework/Versions/A/VideoToolbox
/System/Library/Frameworks/AudioToolbox.framework/Versions/A/AudioToolbox
/System/Library/Frameworks/VideoDecodeAcceleration.framework/Versions/A/VideoDecodeAcceleration
/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics
/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices
/System/Library/Frameworks/CoreAudio.framework/Versions/A/CoreAudio
/System/Library/Frameworks/AudioUnit.framework/Versions/A/AudioUnit
/usr/local/lib/libdvdnav.4.dylib
/usr/lib/libSystem.B.dylib
/usr/local/lib/libdvdread.4.dylib
/usr/local/lib/libdvdcss.2.dylib
/usr/local/lib/liblcms2.2.dylib
/usr/local/lib/libuchardet.0.dylib
/usr/lib/libiconv.2.dylib
/usr/lib/libssl.0.9.8.dylib
/usr/lib/libcrypto.0.9.8.dylib
/usr/lib/libz.1.dylib
/usr/local/lib/libfdk-aac.1.dylib
/usr/local/lib/liblzma.5.dylib
/usr/lib/libbz2.1.0.dylib
/usr/local/lib/libbluray.1.dylib
/usr/local/lib/libxml2.2.dylib
/usr/local/lib/libfreetype.6.dylib
/usr/local/lib/libfontconfig.1.dylib
/usr/local/lib/liblua5.2.4.dylib
/usr/local/lib/libfribidi.0.dylib
/usr/local/lib/libpng16.16.dylib
/usr/local/lib/libharfbuzz.0.dylib
/usr/local/lib/libglib-2.0.0.dylib
/usr/local/lib/libintl.8.dylib
/usr/local/lib/libpcre.1.dylib
/usr/local/lib/libjpeg.9.dylib
/usr/local/lib/libenca.0.dylib
/usr/lib/libobjc.A.dylib
/System/Library/Frameworks/CoreImage.framework/Versions/A/CoreImage
/System/Library/Frameworks/CoreText.framework/Versions/A/CoreText
```

**Comment by ajh17** [comment]
I got the same error today while building. Seems like there's some mismatch between python2 and python3 in the scripts. First of all, this doesn't work in python3: 

```
Creating Mac OS X application bundle (version: git-50991fac8)...
> copying bundle skeleton
> copying binary
> generating Info.plist
> bundling dependencies
Traceback (most recent call last):
  File "TOOLS/dylib-unhell.py", line 94, in <module>
    main()
  File "TOOLS/dylib-unhell.py", line 89, in main
    libs = libraries(binary)
  File "TOOLS/dylib-unhell.py", line 38, in libraries
    result[objfile] = set(libs_list)
  File "TOOLS/dylib-unhell.py", line 15, in is_user_lib
    return not sys_re.match(libname) and \
TypeError: cannot use a string pattern on a bytes-like object
done.
```

When I manually added `.decode()` to libname references, I get the same error as the OP. I then tried to manually change the shebangs to invoke python2 explicitly, which caused the scripts to run without errors and I was able to get a working OS X app.

**Comment by Akemi** [comment]
i fixed this with #5606 hopefully. feel free to test.

**Comment by Akemi** [comment]
fixed via 8ecd404.
